walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -100,26 +102,27 @@
   */
   private ArrayList<TimeRange> mergeTimeRangesThatIntersect(ArrayList<TimeRange> optional, ArrayList<TimeRange> mandatory) {
     ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
-
-    ArrayList<Integer> startValues = new ArrayList<>();
-    for (TimeRange mandatoryTimeRange: mandatory) {
-      startValues.add(mandatoryTimeRange.start());
-    }
+    int mandatoryIndex = 0;
 
     for (TimeRange optionalTimeRange: optional) {
```

Comment:
Set intersection always has to iterate completely through one of the two sets, the one in the outer loop. But sometimes, in the best-case, it doesn't have to iterate through the inner-loop, so best-case complexity is `O(n)` vs. worst-case `O(n + m)`.

You can hyper optimize by putting the shorter set in the outer loop. You can see how this shortcuts the case where your outer-loop set was [1, 2, 3] and the inner loop was [4, 5, 6, 7, 8, 9, 10]

Reply:
I tried this after reading your comment before I realized the runtime is always O(n + m) no matter the order of the loops.

Let's say our smaller array is that of the mandatory time range with available times [0-1440] (Whole day), let's also say my optional times are [0-510], [540-1440]. In this case we should return [0-510], [540-1440] as that's where all optional and mandatory attendees can attend. So either way we order them we had to iterate through both elements of the optional. 

Having said that thanks to this I realized the same is true when the shorter one is the optional, **there can be multiple correct time ranges within one**.  A scenario I hadn't thought about.

Wrote test case. Runtime is still O( n + m ) I believe

Reply:
Oh yeah, you are correct, I was thinking about set intersection, but you're actually doing set union, which much iterate to the end of both sets. 

Nice catch on the test case!